File: morph/ir/css_registry.py
"""Central registry of all CSS properties — single source of truth.

Each entry maps a CSS property to:
  - feature: the MORPH_FEATURE_* flag (None = always included)
  - type: value type for parsing/validation
  - constraints: property inter-dependencies (enables/disables)
"""

CSS_REGISTRY = {
    # ── Base (always present) ──────────────────────────────
    "width":                  {"feature": None,  "type": "px"},
    "height":                 {"feature": None,  "type": "px"},
    "max-width":              {"feature": None,  "type": "px"},
    "background-color":       {"feature": None,  "type": "color"},
    "color":                  {"feature": None,  "type": "color"},
    "border-radius":          {"feature": "radius", "type": "px"},
    "font-size":              {"feature": None,  "type": "px"},
    "font-weight":            {"feature": "bold", "type": "keyword"},
    "text-align":             {"feature": None,  "type": "keyword"},
    "margin":                 {"feature": None,  "type": "spacing"},
    "margin-top":             {"feature": None,  "type": "px"},
    "margin-right":           {"feature": None,  "type": "px"},
    "margin-bottom":          {"feature": None,  "type": "px"},
    "margin-left":            {"feature": None,  "type": "px"},
    "padding":                {"feature": None,  "type": "spacing"},
    "padding-top":            {"feature": None,  "type": "px"},
    "padding-right":          {"feature": None,  "type": "px"},
    "padding-bottom":         {"feature": None,  "type": "px"},
    "padding-left":           {"feature": None,  "type": "px"},
    "overflow":               {"feature": None,  "type": "keyword"},
    "display":                {"feature": None,  "type": "keyword",
                               "values": {"flex": {"enables": ["flex-direction", "justify-content", "align-items", "gap", "flex-wrap", "flex-grow", "flex-shrink", "flex-basis"]}}},
    "box-sizing":             {"feature": None,  "type": "keyword"},
    "flex-direction":         {"feature": "flex", "type": "keyword"},
    "position":               {"feature": None,  "type": "keyword",
                               "values": {"absolute": {"enables": ["left", "right", "top", "bottom"]},
                                          "relative": {"enables": ["left", "right", "top", "bottom"]}}},

    # ── Flex ───────────────────────────────────────────────
    "justify-content":        {"feature": "flex", "type": "keyword"},
    "align-items":            {"feature": "flex", "type": "keyword"},
    "flex-wrap":              {"feature": "flex", "type": "keyword"},
    "flex-grow":              {"feature": "flex", "type": "number"},
    "flex-shrink":            {"feature": "flex", "type": "number"},
    "flex-basis":             {"feature": "flex", "type": "keyword"},
    "gap":                    {"feature": "flex", "type": "px"},

    # ── Position ───────────────────────────────────────────
    "left":                   {"feature": "position", "type": "px"},
    "right":                  {"feature": "position", "type": "px"},
    "top":                    {"feature": "position", "type": "px"},
    "bottom":                 {"feature": "position", "type": "px"},

    # ── Scrollbar ──────────────────────────────────────────
    "scrollbar-width":         {"feature": "scroll", "type": "px"},
    "scrollbar-track-color":   {"feature": "scroll", "type": "color"},
    "scrollbar-thumb-color":   {"feature": "scroll", "type": "color"},
    "scrollbar-border-radius": {"feature": "scroll", "type": "px"},

    # ── Interaction ────────────────────────────────────────
    "cursor":                 {"feature": "cursor", "type": "keyword"},

    # ── Border ─────────────────────────────────────────────
    "border-width":           {"feature": "border", "type": "px"},
    "border-color":           {"feature": "border", "type": "color"},
    "border-style":           {"feature": "border", "type": "keyword"},
    "border":                 {"feature": "border", "type": "shorthand"},
}


def feature_for_property(prop: str) -> str | None:
    """Return the feature flag for a CSS property, or None if always included."""
    entry = CSS_REGISTRY.get(prop)
    return entry["feature"] if entry else None


def properties_for_feature(feature: str) -> list[str]:
    """Return all CSS properties that belong to a given feature."""
    return [p for p, e in CSS_REGISTRY.items() if e["feature"] == feature]


def dependencies(prop: str, value: str) -> list[str]:
    """Return properties enabled by setting `prop` to `value`."""
    entry = CSS_REGISTRY.get(prop)
    if entry and "values" in entry:
        val_entry = entry["values"].get(value)
        if val_entry:
            return val_entry.get("enables", [])
    return []

File: crates/morph-ir/src/css_registry.rs
//! CSS property registry — single source of truth for supported properties
//! Mirrors Python's `morph/ir/css_registry.py` and `morph/style/properties.py`

pub static KNOWN_PROPERTIES: &[&str] = &[
    "display", "position", "top", "right", "bottom", "left", "z-index",
    "float", "clear", "box-sizing", "overflow", "overflow-x", "overflow-y",
    "flex", "flex-direction", "flex-wrap", "flex-flow", "justify-content",
    "align-items", "align-content", "align-self", "flex-grow", "flex-shrink",
    "flex-basis", "gap", "row-gap", "column-gap", "order",
    "width", "height", "min-width", "max-width", "min-height", "max-height",
    "margin", "margin-top", "margin-right", "margin-bottom", "margin-left",
    "padding", "padding-top", "padding-right", "padding-bottom", "padding-left",
    "color", "background-color", "background", "font-size", "font-weight",
    "font-family", "line-height", "text-align", "text-decoration", "text-transform",
    "letter-spacing", "white-space", "word-break",
    "border", "border-width", "border-color", "border-style", "border-radius",
    "border-top", "border-right", "border-bottom", "border-left",
    "opacity", "visibility", "cursor", "box-shadow", "transform", "transform-origin",
    "transition", "transition-duration", "transition-timing-function", "transition-delay",
    "animation", "animation-duration", "animation-timing-function", "animation-delay",
    "animation-iteration-count", "animation-direction", "animation-fill-mode",
    "scrollbar-width", "scrollbar-color",
];

pub static CSS_TO_IR: &[(&str, &str)] = &[
    ("background-color", "bg_color"),
    ("color", "color"),
    ("width", "width"),
    ("height", "height"),
    ("min-width", "min_width"),
    ("max-width", "max_width"),
    ("min-height", "min_height"),
    ("max-height", "max_height"),
    ("margin", "margin"),
    ("padding", "padding"),
    ("border-radius", "border_radius"),
    ("font-size", "font_size"),
    ("font-weight", "font_weight"),
    ("text-align", "text_align"),
    ("display", "display"),
    ("flex-direction", "flex_dir"),
    ("flex-grow", "flex_grow"),
    ("flex-shrink", "flex_shrink"),
    ("flex-basis", "flex_basis"),
    ("gap", "gap"),
    ("position", "position"),
    ("left", "left"),
    ("right", "right"),
    ("top", "top"),
    ("bottom", "bottom"),
    ("justify-content", "justify_content"),
    ("align-items", "align_items"),
    ("flex-wrap", "flex_wrap"),
    ("cursor", "cursor"),
    ("overflow", "overflow"),
    ("border-width", "border_width"),
    ("border-color", "border_color"),
    ("border-style", "border_style"),
    ("box-sizing", "box_sizing"),
    ("z-index", "z_index"),
    ("opacity", "opacity"),
    ("transform", "transform"),
    ("transform-origin", "transform_origin"),
];

pub fn is_known_property(prop: &str) -> bool {
    KNOWN_PROPERTIES.contains(&prop)
}

pub fn property_feature(prop: &str) -> Option<&'static str> {
    match prop {
        "display" | "flex-direction" | "justify-content" | "align-items" | "gap" | "flex-wrap" | "flex-grow" | "flex-shrink" => Some("flex"),
        "border-radius" => Some("radius"),
        "opacity" => Some("opacity"),
        "transform" | "transform-origin" => Some("transform"),
        "animation" | "animation-duration" => Some("animation"),
        "scrollbar-width" | "scrollbar-color" => Some("scrollbar"),
        "position" | "top" | "right" | "bottom" | "left" | "z-index" => Some("positioning"),
        _ => None,
    }
}
